File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/model/PolarExerciseSession.kt
// Copyright © 2025 Polar Electro Oy. All rights reserved.
package com.polar.sdk.api.model

import com.polar.shared.sdk.PolarExerciseSportProfileName
import java.time.LocalDateTime

/**
 * Represents a live exercise session on a Polar device.
 */
class PolarExerciseSession {

    /**
     * Supported sport profiles.
     */
    enum class SportProfile(val id: Int) {
        UNKNOWN(0),
        RUNNING(1),
        CYCLING(2),
        OTHER_OUTDOOR(16);

        companion object {
            /**
             * Resolve [SportProfile] from integer id.
             * Falls back to [UNKNOWN] if no match is found.
             */
            fun fromId(id: Int): SportProfile {
                return when (PolarExerciseSportProfileName.fromId(id)) {
                    PolarExerciseSportProfileName.UNKNOWN -> UNKNOWN
                    PolarExerciseSportProfileName.RUNNING -> RUNNING
                    PolarExerciseSportProfileName.CYCLING -> CYCLING
                    PolarExerciseSportProfileName.OTHER_OUTDOOR -> OTHER_OUTDOOR
                }
            }
        }
    }

    /**
     * Status of an exercise session.
     */
    enum class ExerciseStatus {
        NOT_STARTED,
        IN_PROGRESS,
        PAUSED,
        STOPPED,
        SYNC_REQUIRED
    }

    /**
     * High-level info of current session state.
     */
    data class ExerciseInfo(
        val status: ExerciseStatus,
        val sportProfile: SportProfile,
        val startTime: LocalDateTime? = null
    )
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/model/PolarExerciseSession.swift
//  Copyright © 2025 Polar. All rights reserved.

import Foundation
#if canImport(PolarBleSdkShared)
import PolarBleSdkShared
#endif

/// Represents a live exercise session on a Polar device.
public enum PolarExerciseSession {

    /// Supported sport profiles.
    public enum SportProfile: Int {
        case unknown      = 0
        case running      = 1
        case cycling      = 2
        case otherOutdoor = 16

        public static func from(id: Int) -> Self {
            #if canImport(PolarBleSdkShared)
            guard id >= Int(Int32.min) && id <= Int(Int32.max) else {
                return .unknown
            }
            switch PolarIosSharedBridge.shared.exerciseSportProfileName(id: Int32(id)) {
            case "RUNNING":
                return .running
            case "CYCLING":
                return .cycling
            case "OTHER_OUTDOOR":
                return .otherOutdoor
            default:
                return .unknown
            }
            #else
            return Self(rawValue: id) ?? .unknown
            #endif
        }

        public var displayName: String {
            switch self {
            case .running:      return "Running"
            case .cycling:      return "Cycling"
            case .otherOutdoor: return "Other outdoor"
            case .unknown:      return "Unknown"
            }
        }
    }

    /// Status of an exercise session.
    public enum ExerciseStatus {
        case notStarted
        case inProgress
        case paused
        case stopped
        case syncRequired
    }

    /// High-level info of current session state.
    public struct ExerciseInfo: Equatable, CustomStringConvertible {
        public let status: ExerciseStatus
        public let sportProfile: SportProfile
        public let startTime: Date?

        public init(status: ExerciseStatus, sportProfile: SportProfile, startTime: Date? = nil) {
            self.status = status
            self.sportProfile = sportProfile
            self.startTime = startTime
        }

        public var description: String {
            "ExerciseInfo(status: \(status), sport: \(sportProfile), startTime: \(startTime?.description ?? "nil"))"
        }
    }
}

extension PolarExerciseSession.SportProfile: CaseIterable {
    public static var allCases: [Self] { [.running, .cycling, .otherOutdoor] }
}
